core/arch/arm/kernel/secure_partition.c

```diff
@@ -0,0 +1,772 @@
+				       unsigned long flags)
+{
+	const struct tee_file_operations *fops = NULL;
+	TEE_Result res = TEE_ERROR_BAD_STATE;
+	struct tee_ta_session *sess = NULL;
+	struct tee_file_handle *fh = NULL;
+	struct sec_part_ctx *spc = NULL;
+	struct tee_pobj *po = NULL;
+	size_t file_size = 0;
+	size_t read_len = 0;
+
+	fops = tee_svc_storage_file_ops(storage_id);
+	if (!fops)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (obj_id_len > TEE_OBJECT_ID_MAX_LEN)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = tee_ta_get_current_session(&sess);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	spc = to_sec_part_ctx(sess->ctx);
+	res = tee_mmu_check_access_rights(&spc->uctx,
+					  TEE_MEMORY_ACCESS_WRITE |
+					  TEE_MEMORY_ACCESS_ANY_OWNER,
+					  (uaddr_t)data, len);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = tee_pobj_get(&sess->ctx->uuid, obj_id, obj_id_len, flags,
+			   false, fops, &po);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = po->fops->open(po, &file_size, &fh);
+	if (res != TEE_SUCCESS)
+		goto err_release;
+
+	read_len = len;
+	res = po->fops->read(fh, offset, data, &read_len);
+	if (res != TEE_SUCCESS) {
+		if (res == TEE_ERROR_CORRUPT_OBJECT) {
+			EMSG("Object corrupt");
+			po->fops->remove(po);
+		}
+		goto err_close;
+	}
+
+	/* make sure we read the entire requested length */
+	if (len != read_len) {
+		res = TEE_ERROR_CORRUPT_OBJECT;
+		goto err_close;
+	}
+
+err_close:
+	po->fops->close(&fh);
+err_release:
+	tee_pobj_release(po);
+
+	return res;
+}
+
+/*
+ * Combined write from secure partition, this will create/open, write and
+ * close the fh
+ */
+static TEE_Result sec_storage_obj_write(unsigned long storage_id, char *obj_id,
+					unsigned long obj_id_len, void *data,
+					unsigned long len, unsigned long offset,
+					unsigned long flags)
+
+{
+	const struct tee_file_operations *fops;
+	struct tee_ta_session *sess = NULL;
+	struct tee_file_handle *fh = NULL;
+	struct sec_part_ctx *spc = NULL;
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_pobj *po = NULL;
+
+	fops = tee_svc_storage_file_ops(storage_id);
+	if (!fops)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (obj_id_len > TEE_OBJECT_ID_MAX_LEN)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = tee_ta_get_current_session(&sess);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	spc = to_sec_part_ctx(sess->ctx);
+	res = tee_mmu_check_access_rights(&spc->uctx,
+					  TEE_MEMORY_ACCESS_READ |
+					  TEE_MEMORY_ACCESS_ANY_OWNER,
+					  (uaddr_t)data, len);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = tee_pobj_get(&sess->ctx->uuid, obj_id, obj_id_len, flags,
+			   false, fops, &po);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = po->fops->open(po, NULL, &fh);
+	if (res == TEE_ERROR_ITEM_NOT_FOUND) {
+		res = po->fops->create(po, false, NULL, 0, NULL, 0, NULL, 0,
+				       &fh);
+		if (res != TEE_SUCCESS)
+			goto err_release;
+	}
+
+	res = po->fops->write(fh, offset, data, len);
+	if (res != TEE_SUCCESS)
+		goto err_close;
```

Comment:
can remove the 2 above line.

Reply:
True, but it will be optimized out anyway and imho it's easier to read. Mind keeping it?

Reply:
Oops you are right, I'll fix that